metals/src/main/scala/scala/meta/internal/metals/ImplementationProvider.scala

```diff
@@ -0,0 +1,90 @@
+package scala.meta.internal.metals
+import org.eclipse.lsp4j.Location
+import org.eclipse.lsp4j.TextDocumentPositionParams
+import scala.meta.internal.mtags.Semanticdbs
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+import java.nio.file.Paths
+import scala.meta.internal.semanticdb.TextDocuments
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.semanticdb.ClassSignature
+import scala.meta.internal.semanticdb.TypeRef
+import scala.meta.internal.semanticdb.Signature
+import scala.meta.internal.semanticdb.TextDocument
+import java.util.concurrent.ConcurrentHashMap
+
+final class ImplementationProvider(
+    semanticdbs: Semanticdbs,
+    workspace: AbsolutePath,
+    definitionProvider: DefinitionProvider
+) {
+  private val implementations =
+    new ConcurrentHashMap[String, Set[ClassLocation]]
+
+  def clear(): Unit = {
+    implementations.clear()
+  }
+
+  def implementations(params: TextDocumentPositionParams): List[Location] = {
+    val source = params.getTextDocument.getUri.toAbsolutePath
+    val result = semanticdbs.textDocument(source)
+    for {
+      doc <- result.documentIncludingStale.toList
+      positionOccurrence = definitionProvider.positionOccurrence(
+        source,
+        params,
+        doc
+      )
+      occ <- positionOccurrence.occurrence.toList
+      impl <- findImplementation(occ.symbol)
```

Comment:
The `impl.uri` file may no longer exist since the index doesn't clean up stale data.

Reply:
Fixed and tested.